### ISSUE TITLE
Tensorflow calculates NAN gradients

### ISSUE THREAD

**Issue by farrell236** [geomstats/geomstats#160]
To test the correctness of gradients, I modified and ran ```loss_and_gradient_se3.py```. The following modification has been made in main:

```
    y_pred = SE3.random_uniform(10)
    y_true = SE3.random_uniform(10)

    with tf.GradientTape() as g:
        g.watch(y_pred)
        loss_rot_vec = loss(y_pred, y_true)
        grad_rot_vec = grad(y_pred, y_true)
        dy_dx = g.gradient(loss_rot_vec, y_pred)
        print('The loss between the rotation vectors is: {}'.format(
            loss_rot_vec))
        print('The riemannian gradient is: {}'.format(
            grad_rot_vec))
        print('Tensorflow calculated gradient: {}'.format(
            dy_dx))
```

This is the output I receive:

```
/vol/medic01/users/bh1511/_venv/geomstats-py3/bin/python3.6 /Applications/PyCharm.app/Contents/helpers/pydev/pydevd.py --cmd-line --multiproc --qt-support=auto --client 127.0.0.1 --port 57095 --file /vol/medic01/users/bh1511/_build/geomstats-farrell/examples/loss_and_gradient_se3.py
pydev debugger: process 5100 is connecting

Connected to pydev debugger (build 182.4505.26)
/vol/medic01/users/bh1511/_venv/geomstats-py3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: compiletime version 3.5 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.6
  return f(*args, **kwds)
Using tensorflow backend
2018-11-15 10:02:23.670933: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
The loss between the rotation vectors is: [[7.0383186]
 [3.8544152]
 [3.9183326]
 [3.0725262]
 [2.9405794]
 [7.415542 ]
 [3.8067136]
 [4.2343645]
 [1.4367774]
 [2.0562656]]
The riemannian gradient is: [[ 3.1090195  -1.3207141   2.7089927   1.1367302   1.115068   -1.3076192 ]
 [-1.6554008   0.7796661  -1.7344122  -1.9494474   2.2295787  -0.0385334 ]
 [ 0.45250762 -3.1852932   0.6125859  -0.2580671  -0.10605839 -2.07305   ]
 [-0.12945078  2.7257762  -0.27599958 -1.5244842   1.6424325   0.31258366]
 [-0.02796454  1.8669416  -2.3057005  -1.0833997   1.3495882  -0.2439363 ]
 [-0.71234685 -3.1717374   2.6985168   1.6107618   1.4890841  -1.9891183 ]
 [-0.7424381   2.2465038   1.9941958  -1.8383911   0.7192763  -0.49961993]
 [ 0.6298961   0.58670014 -0.8475082  -3.727872   -0.97431326 -0.5386404 ]
 [-0.06864594  0.507145   -1.2685865  -0.06768421  0.7303784  -1.8040254 ]
 [ 1.1138968  -1.6037818  -0.6482454   0.81384236  1.617016    0.65715736]]
Tensorflow calculated gradient: [[        nan         nan         nan  1.1367303   1.1150677  -1.307619  ]
 [        nan         nan         nan -1.9494474   2.2295775  -0.03853333]
 [        nan         nan         nan -0.2580667  -0.10605845 -2.07305   ]
 [        nan         nan         nan -1.5244837   1.6424323   0.31258366]
 [        nan         nan         nan -1.0834      1.3495883  -0.24393654]
 [        nan         nan         nan  1.6107624   1.489084   -1.9891189 ]
 [        nan         nan         nan -1.8383911   0.7192764  -0.49961996]
 [        nan         nan         nan -3.7278717  -0.974313   -0.53864086]
 [        nan         nan         nan -0.06768411  0.7303788  -1.8040257 ]
 [        nan         nan         nan  0.8138422   1.6170154   0.6571572 ]]
```

Using tensorflow's autograd, the gradient for rotations are nan's. I assume because in the functions, there are checks for if close to pi or 0, somewhere along the path, the gradient is exploding, likely to be a divide by 0 problem. Will need to debug this...

**Comment by johmathe** [comment]
Looking at it in details - stay tuned

**Comment by johmathe** [comment]
Can't reproduce with your code - can you push a branch with your changes please?

**Comment by farrell236** [comment]
johan, ive pushed it here: https://github.com/farrell236/geomstats/tree/tf-nan-gradient-debug

**Comment by farrell236** [comment]
@johmathe, you seem to be using an outdated version of geomstats, this is [line 213](https://github.com/geomstats/geomstats/blob/99899b9548972e64aad49c6a5051033e29e2fd22/geomstats/special_euclidean_group.py#L213) of special_euclidean_group.py

**Comment by johmathe** [comment]
I used the version of geomstats that is in the branch you pointed me to. Can you update the branch with the same version of geomstats you're using?

**Comment by farrell236** [comment]
Are you sure? This is the special_eucidean_group.py file in my branch: [https://github.com/farrell236/geomstats/blob/tf-nan-gradient-debug/geomstats/special_euclidean_group.py](https://github.com/farrell236/geomstats/blob/tf-nan-gradient-debug/geomstats/special_euclidean_group.py)

it does not match line213 in your debug message.

**Comment by johmathe** [comment]
ok I managed to reproduce :)

**Comment by farrell236** [comment]
This issue is fixed with [PR176](https://github.com/geomstats/geomstats/pull/176) 

run script:
```
import os
os.environ['GEOMSTATS_BACKEND'] = 'tensorflow'  # NOQA

import tensorflow as tf
tf.enable_eager_execution()

import numpy as np
import geomstats.lie_group as lie_group
from geomstats.special_euclidean_group import SpecialEuclideanGroup

np.set_printoptions(linewidth=200)

# Reproducable results
np.random.seed(0)

n_vecs = 10

SE3_GROUP   = SpecialEuclideanGroup(3, epsilon=1e-4)
SE3_GROUP_0 = SpecialEuclideanGroup(3)
metric      = SE3_GROUP.left_canonical_metric


y_pred = tf.constant(np.random.rand(n_vecs, 6).astype('float32'))
y_true = tf.constant(np.random.rand(n_vecs, 6).astype('float32'))

with tf.GradientTape(persistent=True) as g:
    g.watch(y_pred)
    loss      = lie_group.loss(y_pred, y_true, SE3_GROUP, metric)
    grad_true = lie_group.grad(y_pred, y_true, SE3_GROUP_0, metric)

grad_tf = g.gradient(loss, y_pred)
print('tensorflow gradients:\n', grad_tf)
print('calculated gradients:\n', grad_true)
print('difference:\n', grad_tf-grad_true)
```

output:
```
/vol/medic01/users/bh1511/_venv/geomstats-py3/bin/python3.6 /Applications/PyCharm.app/Contents/helpers/pydev/pydevconsole.py --mode=client --port=51203
import sys; print('Python %s on %s' % (sys.version, sys.platform))
sys.path.extend(['/vol/medic01/users/bh1511/_build/geomstats-farrell'])
Python 3.6.3 |Intel Corporation| (default, May  3 2018, 23:25:54) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.0.1 -- An enhanced Interactive Python. Type '?' for help.
PyDev console: using IPython 7.0.1
Python 3.6.3 |Intel Corporation| (default, May  3 2018, 23:25:54) 
[GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)] on darwin
runfile('/vol/medic01/users/bh1511/_build/geomstats-farrell/tests/bh_test_tf2.py', wdir='/vol/medic01/users/bh1511/_build/geomstats-farrell/tests')
/vol/medic01/users/bh1511/_venv/geomstats-py3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: compiletime version 3.5 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.6
  return f(*args, **kwds)
Using tensorflow backend
2018-11-30 16:55:44.256013: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
tensorflow gradients:
 tf.Tensor(
[[ 0.7794121   1.1961875  -0.06948584  0.81294453  0.45316723  0.5536636 ]
 [-0.69729364  1.4173661   0.21019858  0.57364213 -0.37031016  0.12049365]
 [-0.80467904  0.4822244  -1.2221593   0.09562665 -0.5257256   1.425117  ]
 [ 1.003707    1.5240505   1.3595973   0.76962817  0.79352975  0.17475086]
 [-0.8880925   0.69913363 -0.7559712   1.702576   -0.10869864 -1.029675  ]
 [-0.09832662  0.21643913  0.62696344 -0.29603294 -0.54107606  0.86805546]
 [ 0.06865987  1.110144    0.23793712  1.353436   -0.6377071   0.33405042]
 [-0.15951686 -1.707704    0.7036152   0.18854514 -0.76333934 -0.8860116 ]
 [ 0.15188618 -1.1487913   0.1921737  -0.816476    0.5785309  -0.39112929]
 [-1.1606852  -0.45982727 -0.5063167  -0.65677613 -0.8301717  -0.89581335]], shape=(10, 6), dtype=float32)
calculated gradients:
 tf.Tensor(
[[ 0.76598585  1.229558   -0.16529617  0.8127196   0.45332846  0.553708  ]
 [-0.80276626  1.5810819   0.14563479  0.5730106  -0.3697986   0.12050432]
 [-0.8595346   0.5488441  -1.2293453   0.09584883 -0.5250818   1.4237987 ]
 [ 0.9178593   1.53511     1.330125    0.7694223   0.79329824  0.17516279]
 [-0.8923641   0.7174605  -0.7615303   1.7015893  -0.10855845 -1.0288775 ]
 [-0.11131512  0.20898023  0.645187   -0.29577032 -0.54070556  0.8673688 ]
 [ 0.01711899  1.1807405   0.18895364  1.3523666  -0.6366438   0.33402926]
 [-0.1887154  -1.7141953   0.7449368   0.18806021 -0.76299125 -0.88572156]
 [ 0.16241625 -1.1805391   0.21431005 -0.8160132   0.57795936 -0.39101624]
 [-1.1684107  -0.45754045 -0.5006553  -0.65681845 -0.8300786  -0.8957907 ]], shape=(10, 6), dtype=float32)
difference:
 tf.Tensor(
[[ 1.34262443e-02 -3.33704948e-02  9.58103314e-02  2.24947929e-04 -1.61230564e-04 -4.44054604e-05]
 [ 1.05472624e-01 -1.63715720e-01  6.45637959e-02  6.31511211e-04 -5.11556864e-04 -1.06692314e-05]
 [ 5.48555851e-02 -6.66196942e-02  7.18605518e-03 -2.22176313e-04 -6.43789768e-04  1.31833553e-03]
 [ 8.58477354e-02 -1.10595226e-02  2.94723511e-02  2.05874443e-04  2.31504440e-04 -4.11927700e-04]
 [ 4.27156687e-03 -1.83268785e-02  5.55908680e-03  9.86695290e-04 -1.40197575e-04 -7.97510147e-04]
 [ 1.29885003e-02  7.45889544e-03 -1.82235837e-02 -2.62618065e-04 -3.70502472e-04  6.86645508e-04]
 [ 5.15408814e-02 -7.05964565e-02  4.89834845e-02  1.06942654e-03 -1.06328726e-03  2.11596489e-05]
 [ 2.91985422e-02  6.49130344e-03 -4.13216352e-02  4.84928489e-04 -3.48091125e-04 -2.90036201e-04]
 [-1.05300695e-02  3.17478180e-02 -2.21363455e-02 -4.62770462e-04  5.71548939e-04 -1.13040209e-04]
 [ 7.72547722e-03 -2.28682160e-03 -5.66142797e-03  4.23192978e-05 -9.31024551e-05 -2.26497650e-05]], shape=(10, 6), dtype=float32)

```

**Comment by farrell236** [comment]
This issue can now be closed, will update with example.

Run script:
```
import os
os.environ['GEOMSTATS_BACKEND'] = 'tensorflow'  # NOQA

import tensorflow as tf

from tensorflow.contrib.slim.python.slim.nets import inception
from tensorflow.contrib.slim.python.slim.nets import alexnet

import numpy as np
import geomstats.lie_group as lie_group
from geomstats.special_euclidean_group import SpecialEuclideanGroup

np.set_printoptions(linewidth=200)

# Reproducable results
# np.random.seed(0)
epsilon = np.finfo(np.float32).eps # 1.1920929e-07

n_vecs = 4

SE3_GROUP   = SpecialEuclideanGroup(3, epsilon=epsilon)
metric      = SE3_GROUP.left_canonical_metric

image       = tf.placeholder(tf.float32, shape=[n_vecs, 224, 224, 3])
y_true      = tf.placeholder(tf.float32, shape=[n_vecs, 6])

_image      = np.random.rand(n_vecs, 224, 224, 3).astype('float32')
_y_true     = np.random.rand(n_vecs, 6).astype('float32')

# net         = inception.inception_v1(image, num_classes=6)
net         = alexnet.alexnet_v2(image, num_classes=6)

loss        = lie_group.loss(net[0], y_true, SE3_GROUP, metric)

optimizer   = tf.train.AdamOptimizer(learning_rate=0.001).minimize(loss)

init        = tf.global_variables_initializer()

grad_tf = tf.gradients(loss, net[0])
grad_true     = lie_group.grad(net[0], y_true, SE3_GROUP, metric)


with tf.Session() as sess:

    sess.run(init)

    for i in range(0, 100):
        print('Iteration:', i)
        _, _loss, _y_pred, _grad_tf, _grad_true = sess.run([optimizer, loss, net[0], grad_tf, grad_true],
                                                  feed_dict={image: _image, y_true: _y_true})
        print('loss:\n', _loss)
        print('y_pred:\n', _y_pred)
        print('y_true:\n', _y_true)
        print('grad_tensorflow:\n', _grad_tf[0])
        print('grad_riemannian:\n', _grad_true)
        print('grad_diff:\n', np.abs(_grad_true - _grad_tf[0]))

```

Log Output: [output_log.txt](https://github.com/geomstats/geomstats/files/2640541/output_log.txt)
